### PR TITLE
Enchance transformer model

### PR DESCRIPTION
**Pull request by peterprospl12**

### CODE REVIEW

**Comment by tomaszsankowski** [comment]
Didn't we decided to not to use padding token, so this PR can be closed?

File: breathing_model/model/transformer/dataset.py [hunk 1]
```diff
@@ -126,27 +126,30 @@ def collate_fn(batch):
...
-        labels: (batch_size, time_frames_padded)
+        spectrograms_batch: [batch_size, 1, n_mels, time_frames_padded]
+        labels_padded: [batch_size, time_frames_padded]
+        padding_mask: [batch_size, time_frames_padded] (bool) True where padding
     """
     spectrograms, labels = zip(*batch)
 
-    # Transform spectrograms: (1, n_mels, time_frames) → (time_frames, n_mels)
-    spectrograms_transposed = [spectrogram.squeeze(0).permute(1, 0) for spectrogram in spectrograms]
+    # Remember original lenghts before padding
```
[the remark this answers is not in the document]
Comment (peterprospl12) [reply breathing_model/model/transformer/dataset.py:135]: done

File: breathing_model/model/transformer/dataset.py [hunk 2]
```diff
@@ -126,27 +126,30 @@ def collate_fn(batch):
...
+    original_lengths = [spec.shape[-1] for spec in spectrograms]
 
-    # (batch_size, time_frames_max, n_mels=128)
-    spectrograms_padded = pad_sequence(spectrograms_transposed,
-                                       batch_first=True,
-                                       padding_value=0.0)
+    # Convert spectrograms from shape (1, n_mels, T) -> (T, n_mels) for pad_sequence,
+    # then pad and permute back to [B, 1, n_mels, T]
```
[the remark this answers is not in the document]
Comment (peterprospl12) [reply breathing_model/model/transformer/dataset.py:139]: done

File: breathing_model/model/transformer/train.py [hunk 3]
```diff
@@ -1,148 +1,239 @@
...
-        loss = criterion(outputs.view(-1, outputs.size(-1)), labels.view(-1))
-        loss.backward()
-        optimizer.step()
 
-        running_loss += loss.item() * inputs.size(0)
-        _, preds = torch.max(outputs, dim=-1)
-        total_frames += labels.numel()
-        correct_frames += (preds == labels).sum().item()
+        # Forward pass: pass src_key_padding_mask (True = padded positions)
+        outputs = model(spectograms_batch, src_key_padding_mask=padding_mask_batch) # [batch_size,
```
[the remark this answers is not in the document]
Comment (peterprospl12) [reply breathing_model/model/transformer/train.py:37]: done

File: breathing_model/model/transformer/train.py [hunk 4]
```diff
@@ -1,148 +1,239 @@
...
 
-    for inputs, labels in train_loader:
-        inputs = inputs.to(device)
-        labels = labels.to(device)
+    total_loss_weighted = 0.0
+    total_valid_frames = 0
+    total_correct_predictions = 0
+
+    for batch_index, (spectograms_batch, labels_batch, padding_mask_batch) in enumerate(data_loader):
+        spectograms_batch = spectograms_batch.to(device)
```
[the remark this answers is not in the document]
Comment (peterprospl12) [reply breathing_model/model/transformer/train.py:27]: done

**Review by tomaszsankowski — APPROVED** [review]
